Add support for a COBALT_param_doc file

theresa-cordero opened the pull request:
Using the file parser from MOM6, add the option to read parameters from COBALT_input and COBALT_override files and write parameters to COBALT_param_doc files. 

With these changes, COBALT_input and COBALT_override files must be listed in the input namelist. As a result all existing experiments will need their namelists updated and COBALT_input files generated. The files can be blank and the default values will be used.  
 ```
&cobalt_input_nml
             parameter_filename = 'INPUT/COBALT_input','INPUT/COBALT_override'
/
```

Descriptions of the parameters have not been added with this PR and some parameters lack units. For the description, the parameter name has been used but should be updated to include a note about what the parameter is used for and a recommended range (or that the parameter not be adjusted). 


As @amoebaliz suggested, here are the needed changes to make to an xml to use the parameter doc file.

1.  In the `<input>` section for each experiment or for the primary experiment, add the new namelist option:
```
<namelist name = cobalt_input_nml >
  parameter_filename = 'INPUT/COBALT_input','INPUT/COBALT_override'
</namelist>
```
2. You can create a `COBALT_input` file in the same location as your `MOM_input` file and link to is in a similar way, for example:
```
<dataSource site="ncrc">$(ESMG_H)/ESMG-configs/Arctic6/mid_12/MOM_input</dataSource>
<dataSource site="ncrc">$(ESMG_H)/ESMG-configs/Arctic6/mid_12/COBALT_input</dataSource>
```
3.  Alternatively you can write a `COBALT_input` file directly from the xml. Toward the end of the `input` section, where the `MOM_override` file is defined in you can add:
```
<csh type='always'>
<![CDATA[
truncate -s 0 $work/INPUT/MOM_override
cat > $work/INPUT/MOM_override << MOM_OVERRIDE_EOF
MOM_OVERRIDE_EOF

truncate -s 0 $work/INPUT/COBALT_input
cat > $work/INPUT/COBALT_input << COBALT_INPUT_EOF
! Put the contents of your COBALT_input file here
COBALT_INPUT_EOF
]]>
</csh>
```
This is how and where you can add a COBALT_override file
```
truncate -s 0 $work/INPUT/COBALT_override
cat > $work/INPUT/COBALT_override << COBALT_OVERRIDE_EOF
! Put the contents of your COBALT_override file here.
! For parameters which are not specified in the COBALT_input file, just set the new value, ex:
! do_case2_mod = True
! For parameters which are specified in the COBALT_input file, specify an override, ex:
! #override do_case2_mod = True
COBALT_OVERRIDE_EOF
```

5. Finally, you can update the post processing section to include the COBALT_parameter_doc files.
```
    <postProcess>
      <csh><![CDATA[
         cd $work
         #Make a directory to trick FRE to pick up and archive in ascii
         mkdir extra.results
         mv *velocity_truncations MOM_parameter_doc* SIS_parameter_doc* COBALT_parameter_doc* seaice.stats* ocean.stats* extra.results/
         cp diag_table extra.results/.
         cp $scriptName extra.results/runscript.csh
         #When the ocean uses a mask_table the ocean_static.nc file produced by the model run has holes in coordinates (geolon,geolat)
         #This causes problems for ferret and python tools for analysis.
         #Copy a non-masked version of ocean_static.nc to be saved as a history file to be used by the analysis scipts .
         #cp $work/INPUT/ocean_static_no_mask_table.nc $work/
      ]]></csh>
      <xi:include xpointer="xpointer(//freInclude[@name='MOM6_postprocess']/postProcess/node())"/>
    </postProcess>
```

Commit:
Add support for a COBALT_param_doc file

Using the file parser from MOM6, add the option to read parameters
from COBALT_input and COBALT_override files and write pramerters to
COBALT_param_doc files. The COBALT_input files must be specific in
the input.nml

Commit:
Replace most calls to g_tracer_add_param

Replace calls to g_tracer_add_param with get_param but do not change deaults.
The descriptions are just the parameter name, and the units are included where
they were documenemted in the code.

yichengt900 commented:
Thanks @theresa-morrison . I have updated the [github actions testing workflow](https://github.com/NOAA-CEFI-Regional-Ocean-Modeling/ocean_BGC/pull/62) to accommodate this PR. Please update your feature branch, and it should be able to pass the RT.

I can confirm that this new functionality works properly. For example, if I set `remin_ramp_scale = 100.0` in COBALT_input and use `COBALT_override` to revert it to the default value, the model results are identical to the reference.

@charliestock and anyone else, it would be great if you could double-check the changes in `generic_tracers/generic_COBALT.F90` and ensure you are comfortable with these updates.

theresa-cordero committed:
Merge pull request #2 from NOAA-CEFI-Regional-Ocean-Modeling/dev/cefi

rebase

Commit:
Add cobalt_param_doc.F90

Using the file parser from MOM6, add the option to read parameters
from COBALT_input and COBALT_override files and write pramerters to
COBALT_param_doc files. The COBALT_input files must be specific in
the input.nml

Commit:
Replace most calls to g_tracer_add_param

Add a call to read the name and loaction of the COBALT param files
from the namelist to generic_COBALT

Replace calls to g_tracer_add_param with get_param but do not change deaults.
The descriptions are just the parameter name, and the units are included where
they were documenemted in the code.

Commit:
Fix typo

Fix error in line continue statement.

Commit:
Add missing description

Add description for RHO_0 and NKML

Commit:
Cange parameters with sperd or spery

Remove unit changes from per-day or per-year to per-second from
the default values to do that calculation outside the get param
call. This should not change the solution.

theresa-cordero commented:
As @amoebaliz suggested, here are the needed changes to make to an xml to use the parameter doc file.

1.  In the `<input>` section for each experiment or for the primary experiment, add the new namelist option:
```
<namelist name = cobalt_input_nml >
  parameter_filename = 'INPUT/COBALT_input','INPUT/COBALT_override'
</namelist>
```
2. You can create a `COBALT_input` file in the same location as your `MOM_input` file and link to is in a similar way, for example:
```
<dataSource site="ncrc">$(ESMG_H)/ESMG-configs/Arctic6/mid_12/MOM_input</dataSource>
<dataSource site="ncrc">$(ESMG_H)/ESMG-configs/Arctic6/mid_12/COBALT_input</dataSource>
```
3.  Alternatively you can write a `COBALT_input` file directly from the xml. Toward the end of the `input` section, where the `MOM_override` file is defined in you can add:
```
<csh type='always'>
<![CDATA[
truncate -s 0 $work/INPUT/MOM_override
cat > $work/INPUT/MOM_override << MOM_OVERRIDE_EOF
MOM_OVERRIDE_EOF

truncate -s 0 $work/INPUT/COBALT_input
cat > $work/INPUT/COBALT_input << COBALT_INPUT_EOF
! Put the contents of your COBALT_input file here
COBALT_INPUT_EOF
]]>
</csh>
```
This is how and where you can add a COBALT_override file
```
truncate -s 0 $work/INPUT/COBALT_override
cat > $work/INPUT/COBALT_override << COBALT_OVERRIDE_EOF
! Put the contents of your COBALT_override file here.
! For parameters which are not specified in the COBALT_input file, just set the new value, ex:
! do_case2_mod = True
! For parameters which are specified in the COBALT_input file, specify an override, ex:
! #override do_case2_mod = True
COBALT_OVERRIDE_EOF
```

5. Finally, you can update the post processing section to include the COBALT_parameter_doc files.
```
    <postProcess>
      <csh><![CDATA[
         cd $work
         #Make a directory to trick FRE to pick up and archive in ascii
         mkdir extra.results
         mv *velocity_truncations MOM_parameter_doc* SIS_parameter_doc* COBALT_parameter_doc* seaice.stats* ocean.stats* extra.results/
         cp diag_table extra.results/.
         cp $scriptName extra.results/runscript.csh
         #When the ocean uses a mask_table the ocean_static.nc file produced by the model run has holes in coordinates (geolon,geolat)
         #This causes problems for ferret and python tools for analysis.
         #Copy a non-masked version of ocean_static.nc to be saved as a history file to be used by the analysis scipts .
         #cp $work/INPUT/ocean_static_no_mask_table.nc $work/
      ]]></csh>
      <xi:include xpointer="xpointer(//freInclude[@name='MOM6_postprocess']/postProcess/node())"/>
    </postProcess>
```

theresa-cordero commented:
With the latest commit, parameters with default values that included a conversion from per-day or per-year ito per-second in the default have been updated so that there default is the per-day or per-year value. 

These value are better understood in the community which will make it easier for users to adjust them. 

In addition, avoiding division in calculation of the default is important for producing accurate values in the parameter doc files. Otherwise, when the rounded values from the parameter doc file are specified rather than the default the difference changes the model results.

This commit did not impact the parameters with calculations in the default that appeared to be related to stoichiometry. These parameters should be clearly labeled in their descriptions possibly with a warning that users should not adjust these values.

theresa-cordero commented on generic_tracers/generic_COBALT.F90:
```diff
@@ -496,456 +508,507 @@ subroutine user_add_params
-    call g_tracer_add_param('k_po4_Sm', phyto(SMALL)%k_po4,  1.0e-8)                  ! mol PO4 kg-1
-    call g_tracer_add_param('k_sio4_Lg',phyto(LARGE)%k_sio4, 2.0e-6)                  ! mol SiO4 kg-1
-    call g_tracer_add_param('k_sio4_Md',phyto(MEDIUM)%k_sio4,1.0e-6)                  ! mol SiO4 kg-1
-    call g_tracer_add_param('k_fe_2_n_Di', phyto(DIAZO)%k_fe_2_n, 12.0e-6 * 106.0 / 16.0)   ! mol Fe mol N-1
```
@charliestock @amoebaliz @andrew-c-ross  
Are these parameter defaults that depend on stoichiometry or is there another reason for the division?

charliestock replied:
Yes - values in the literature are generally given in units of iron to carbon.  Since nitrogen serves as the main currency in COBALT and other BGC models, we covert to iron to nitrogen by using the Redfield ratio of 106 carbons per unit of nitrogen.

This could be handled with a "sperd"-like conversion c2n = 106/16?

theresa-cordero replied:
added `c2n` conversion.

theresa-cordero commented on generic_tracers/generic_COBALT.F90:
```diff
@@ -496,456 +508,507 @@ subroutine user_add_params
-    call g_tracer_add_param('alk_2_n_denit', cobalt%alk_2_n_denit, 552.0/472.0)             ! eq. alk mol NO3-1
-    call g_tracer_add_param('alk_2_nh4_amx', cobalt%alk_2_nh4_amx, 3732.0/8030.0)           ! eq. alk mol NH4-1
-    call g_tracer_add_param('p_2_n_static_Di', phyto(DIAZO)%p_2_n_static,1.0/40.0 )         ! mol P mol N-1
-    call g_tracer_add_param('p_2_n_static_Lg', phyto(LARGE)%p_2_n_static,1.0/14.0 )         ! mol P mol N-1
```
@charliestock @amoebaliz @andrew-c-ross 
Can these 1.0/X parameters be changed to X and then we can do the 1.0/X calculation on the next line? 
Some of the resulting defaults cannot be exactly represented in the parameter file.

yichengt900 replied:
@theresa-morrison, I will let the others answer your concern, but I think that we probably do not want users to realize that they can easily/directly change these `classic stoichiometry` parameters at this moment. Perhaps there is an option in the `get_param` call to do so.

charliestock replied:
I think we a reasonable level of accuracy is fine here (i.e., we could convert to a decimal value if needed).  That said, I agree with @yichengt900 that there are some values that should not be changed.  @theresa-morrison, my understanding from our conversation yesterday was that we could set this with an option not to read external values for certain parameters that should not be changed?

charliestock replied:
I obviously didn't read your next comment.....please see additional thoughts below.

theresa-cordero replied:
added do_not_read=true for these variables.

theresa-cordero commented:
There are two relevant options for `get_param` that we can use here:
`do_not_read` : If present and true, do not read a value for this parameter, although it might be logged.
`scale` : A scaling factor that the parameter is multiplied by before it is returned.

`do_not read` would be good for the stoichiometry parameters that users may want to make sure are consistent, but should not be able to easily (or accidentally) change. 

`scale` is similar to what Xiao suggested in the documentation meeting yesterday, and would be a good option for the time unit conversions that some of the parameters have.

charliestock commented on generic_tracers/cobalt_param_doc.F90:
```diff
@@ -0,0 +1,86 @@
+  !                                                             !! input_filename if input_filename is no listed
+  !                                                             !! in the namelist MOM_input_nml.
+  ! Local variables
+  integer, parameter :: npf = 5 ! Maximum number of parameter files
```
Is there a case where we would need more than one?

charliestock commented on generic_tracers/generic_COBALT.F90:
```diff
@@ -496,456 +508,507 @@ subroutine user_add_params
-    call g_tracer_add_param('z_sed',  cobalt%z_sed, 0.1 )                                    ! m
-    call g_tracer_add_param('k_no3_denit',cobalt%k_no3_denit,1.0e-6)                         ! mol NO3 kg-1
-    call g_tracer_add_param('z_burial',cobalt%z_burial,10.0)                                 ! m
+    call get_param(param_file, "generic_COBALT", "k_o2",             cobalt%k_o2,              "k_o2",             units="mol O2 kg-1   ", default= 8.0e-6)                                     ! mol O2 kg-1
```
Note that some concentrations are currently entered as moles kg-1 (i.e., after division by 1e6), while others are entered in micromoles kg-1 units and then converted.  We should likely uniformly do the latter?

charliestock commented on generic_tracers/generic_COBALT.F90:
```diff
@@ -496,456 +508,507 @@ subroutine user_add_params
+    call get_param(param_file, "generic_COBALT", "kappa_sidet",      cobalt%kappa_sidet,       "kappa_sidet",      units="deg C -1      ", default= 0.063 )                      ! deg C -1
+    call get_param(param_file, "generic_COBALT", "gamma_sidet",      cobalt%gamma_sidet,       "gamma_sidet",      units="s-1           ", default= cobalt%wsink / 1.0e4 )       ! s-1
+    call get_param(param_file, "generic_COBALT", "phi_lith" ,        cobalt%phi_lith,          "phi_lith" ,        units="dimensionless ", default= 0.002)                            ! dimensionless
+    call get_param(param_file, "generic_COBALT", "k_lith",           cobalt%k_lith,            "k_lith",           units="year-1        ", default= 0.5)                            ! year-1
```
Should we require this to be entered in day-1 to prevent too many different time scales and conversions?  There are certainly long time scale processes that are commonly discussed in units of years, so maybe it is OK to have both.  We'd just need to be sure the user has ready access to all the required units.

charliestock commented on generic_tracers/generic_COBALT.F90:
```diff
@@ -496,456 +508,507 @@ subroutine user_add_params
-    call g_tracer_add_param('gamma_srdop',  cobalt%gamma_srdop, 1.0 / (4.0 * spery))           ! s-1
-    call g_tracer_add_param('gamma_sldon',  cobalt%gamma_sldon, 1.0 / (90 * sperd))           ! s-1
-    call g_tracer_add_param('gamma_sldop',  cobalt%gamma_sldop, 1.0 / (90 * sperd))           ! s-1
+    call get_param(param_file, "generic_COBALT", "gamma_srdon",  cobalt%gamma_srdon, "gamma_srdon", units="1/(default*spery)", default= 10.0)
```
The "10" here is 10 years, so the model units "sec-1".  We can eventually make this "day-1" to simplify?

charliestock commented on generic_tracers/generic_COBALT.F90:
```diff
@@ -496,456 +508,507 @@ subroutine user_add_params
-    !call g_tracer_add_param('gamma_nh4amx',  cobalt%gamma_nh4amx, 0.07 / sperd)                      ! s-1
-    call g_tracer_add_param('gamma_nh4amx',  cobalt%gamma_nh4amx, 0.0 / sperd)                       ! s-1
-    call g_tracer_add_param('o2_min_amx', cobalt%o2_min_amx, 4.0e-6 )                                ! mol O2 kg-1
+    call get_param(param_file, "generic_COBALT", "gamma_nitrif",  cobalt%gamma_nitrif, "gamma_nitrif", units="gamma_nitrif /(default*sperd)", default= 30.0)     ! s-1
```
Need to remove Fabian's namelists references here, but we can leave this for the refinement after Theresa's foundational pull request.

charliestock commented:
Thanks again for the great work on this pull request Theresa.  It is an exciting development!  

The flexibility offered by "do_not_read" and "scale" sounds great.  For "scale" it seems that we have a limited number of primary conversions that may make this approach tractable:

1) Time: My inclination is to require all inputs in "days" or "days-1" and then convert to "seconds" or "seconds-1" using sperd = 86400.  There are some parameters that we have converted from "year" or "year-1" in the past, but I don't think there are many of these and it may be easiest to limit the number.  Thoughts?

2) Concentrations: We could have the input unit be micromoles kg-1 and use a uniform conversion micromol2mol = 1e6 to convert?

3) Nutrient ratios: There are a number of ratios that are commonly reported in the literature relative to carbon (i.e., iron to carbon) that are converted into ratios relative to nitrogen for the model.  A conversion c2n = 106.0/16.0 would do the trick for these.

4) Light: Most of the photosynthesis literature reports parameters, such as the initial slope of photosynthesis-irradiance curve or the compensation irradiance in "micromoles quanta" of photosynthetically available radiation.  We use a conversion from Smith and Morel (1974) to convert this to watts: under typical conditions 2.77e18 quanta sec-1 = a watt.  So there are 2.77e18/6.022e17 = 4.6 micromole quanta per sec-1 per Watt (note that 6.022e17 is a micromole).  How about something like micromolQpersec2W = 2.77e18/6.022e17?  Any suggestions on how to shorten?

I may have missed a few, but I think this would cover most of them.

Lastly, I'm wondering when the best time to accept this initial pull request is?  It seems that we'd want to accept the foundation of the pull request ASAP so that we can begin work on filling in parameter details, etc.  Do folks think we should pull things over before "scaling" settings and then work to refine and rescale together?  Or is it worth implementing/testing the rescaling first?  

Have we confirmed that this doesn't change answers for a more complete run like NWA12?

Again - excellent work Theresa and all who have contributed!

Charlie

yichengt900 commented on generic_tracers/generic_COBALT.F90:
```diff
@@ -496,456 +508,507 @@ subroutine user_add_params
+    call get_param(param_file, "generic_COBALT", "irr_mem_dpthresh1",cobalt%irr_mem_dpthresh1, "irr_mem_dpthresh1",units="watts m-2", default=30.0) ! watts m-2
+    call get_param(param_file, "generic_COBALT", "irr_mem_dpthresh2",cobalt%irr_mem_dpthresh2, "irr_mem_dpthresh2",units="watts m-2", default=10.0) ! watts m-2
+    call get_param(param_file, "generic_COBALT", "dpause_max",       cobalt%dpause_max,        "dpause_max",       units="", default=0.0)           ! dimensionless
+    call get_param(param_file, "generic_COBALT", "upswim_chl_thresh",zoo(1)%upswim_chl_thresh, "upswim_chl_thresh",units="", default=0.0) ! dimensionless
+    call get_param(param_file, "generic_COBALT", "upswim_chl_thresh",zoo(2)%upswim_chl_thresh, "upswim_chl_thresh",units="", default=0.0) ! dimensionless
+    call get_param(param_file, "generic_COBALT", "upswim_chl_thresh",zoo(3)%upswim_chl_thresh, "upswim_chl_thresh",units="", default=0.0) ! dimensionless
+    call get_param(param_file, "generic_COBALT", "upswim_I_thresh",  zoo(1)%upswim_I_thresh,   "upswim_I_thresh",  units="", default=0.0)     ! dimensionless
+    call get_param(param_file, "generic_COBALT", "upswim_I_thresh",  zoo(2)%upswim_I_thresh,   "upswim_I_thresh",  units="", default=0.0)     ! dimensionless
+    call get_param(param_file, "generic_COBALT", "upswim_I_thresh",  zoo(3)%upswim_I_thresh,   "upswim_I_thresh",  units="", default=0.0)     ! dimensionless
+    call get_param(param_file, "generic_COBALT", "swim_max",         zoo(1)%swim_max,          "swim_max",         units="m day-1", default=100.0)
+      zoo(1)%swim_max = zoo(1)%swim_max/sperd! max swimming (m sec-1)
+    call get_param(param_file, "generic_COBALT", "swim_max",         zoo(2)%swim_max,          "swim_max",         units="m day-1", default=500.0)
+      zoo(2)%swim_max = zoo(2)%swim_max/sperd! max swimming (m sec-1)
+    call get_param(param_file, "generic_COBALT", "swim_max",         zoo(3)%swim_max,          "swim_max",         units="m day-1", default=2000.0)
+      zoo(3)%swim_max = zoo(3)%swim_max/sperd! max swimming (m sec-1)
```
Since we have 3 zooplankton groups here,  I suggested to change these codes as follows:"
```
    call get_param(param_file, "generic_COBALT", "upswim_chl_thresh1",zoo(1)%upswim_chl_thresh, "upswim_chl_thresh1",units="", default=0.0) ! dimensionless
    call get_param(param_file, "generic_COBALT", "upswim_chl_thresh2",zoo(2)%upswim_chl_thresh, "upswim_chl_thresh2",units="", default=0.0) ! dimensionless
    call get_param(param_file, "generic_COBALT", "upswim_chl_thresh3",zoo(3)%upswim_chl_thresh, "upswim_chl_thresh3",units="", default=0.0) ! dimensionless
    call get_param(param_file, "generic_COBALT", "upswim_I_thresh1",  zoo(1)%upswim_I_thresh,   "upswim_I_thresh1",  units="", default=0.0)     ! dimensionless
    call get_param(param_file, "generic_COBALT", "upswim_I_thresh2",  zoo(2)%upswim_I_thresh,   "upswim_I_thresh2",  units="", default=0.0)     ! dimensionless
    call get_param(param_file, "generic_COBALT", "upswim_I_thresh3",  zoo(3)%upswim_I_thresh,   "upswim_I_thresh3",  units="", default=0.0)     ! dimensionless
    call get_param(param_file, "generic_COBALT", "swim_max1",         zoo(1)%swim_max,          "swim_max1",         units="m day-1", default=100.0)
      zoo(1)%swim_max = zoo(1)%swim_max/sperd! max swimming (m sec-1)
    call get_param(param_file, "generic_COBALT", "swim_max2",         zoo(2)%swim_max,          "swim_max2",         units="m day-1", default=500.0)
      zoo(2)%swim_max = zoo(2)%swim_max/sperd! max swimming (m sec-1)
    call get_param(param_file, "generic_COBALT", "swim_max3",         zoo(3)%swim_max,          "swim_max3",         units="m day-1", default=2000.0)
      zoo(3)%swim_max = zoo(3)%swim_max/sperd! max swimming (m sec-1)
```

charliestock replied:
Good catch YC.  We could also change up "upswim_I_thresh_smz, upswim_I_thresh_mdz, upswim_I_thresh_lgz" for even more clarity.  That said, the upward swimming isn't activated since it didn't help in the way I hoped so we may ultimately want to clean this up in favor of a more complete zooplankton vertical movement later.

yichengt900 commented:
Thanks, @theresa-morrison, for this great work, and thank you, @Charlie, for the detailed and thoughtful comments. I agree with your points and appreciate the clarity on the primary conversions.

I have a minor comment regarding the zooplankton swimming parameters that could help us maintain answers (it's not activate at this moment but the `get_param` call will lock the parameter so it's better to use different parameter names for each groups) in the NWA12 case.

Regarding the `do_case2_mod=false` setting, do we prefer this as the default?

I also support accepting this PR sooner rather than later. We can revisit the rescaling in a separate PR. This would allow other developers to benefit from the enhancements and begin working with the new functionality immediately.

Great work, @theresa-morrison, and thanks to everyone involved!

theresa-cordero commented:
>     1. Time: My inclination is to require all inputs in "days" or "days-1" and then convert to "seconds" or "seconds-1" using sperd = 86400.  There are some parameters that we have converted from "year" or "year-1" in the past, but I don't think there are many of these and it may be easiest to limit the number.  Thoughts?

I agree with this, there are about 8 `spery` conversions. I think it would be fine to keep those around for now.
 
>     2. Concentrations: We could have the input unit be micromoles kg-1 and use a uniform conversion micromol2mol = 1e6 to convert?

I like this because it will make it very clear what the 1e6 is for. 

>     3. Nutrient ratios: There are a number of ratios that are commonly reported in the literature relative to carbon (i.e., iron to carbon) that are converted into ratios relative to nitrogen for the model.  A conversion c2n = 106.0/16.0 would do the trick for these.

I can do this, I will just search for "106.0/16.0" which should get most instances.

>     4. Light: Most of the photosynthesis literature reports parameters, such as the initial slope of photosynthesis-irradiance curve or the compensation irradiance in "micromoles quanta" of photosynthetically available radiation.  We use a conversion from Smith and Morel (1974) to convert this to watts: under typical conditions 2.77e18 quanta sec-1 = a watt.  So there are 2.77e18/6.022e17 = 4.6 micromole quanta per sec-1 per Watt (note that 6.022e17 is a micromole).  How about something like micromolQpersec2W = 2.77e18/6.022e17?  Any suggestions on how to shorten?

I think `micromolQpersec2W` is clear and I can include a reference to Smith and Morel (1974)  and some of the description here where the parameter is defined. These scaling factors will only be in the parameter read statements so it is okay if they are long. 
 
> Lastly, I'm wondering when the best time to accept this initial pull request is? It seems that we'd want to accept the foundation of the pull request ASAP so that we can begin work on filling in parameter details, etc. Do folks think we should pull things over before "scaling" settings and then work to refine and rescale together? Or is it worth implementing/testing the rescaling first?
> 
> Have we confirmed that this doesn't change answers for a more complete run like NWA12?

I can make the changes and do some basic testing on my workstation today. I have done short tests with the NWA12 but I would recommend we run ~1 year in case there are parameters that aren't active in the first month or few days of the year. I have only been looking at the ocean stats file, but we should look at the COBALT values in the standard out file as well.

I also agree with @yichengt900  that we can put the rescaling in a separate PR to expedite this. And I will change `do_case2_mod=true` back to the default.

charliestock commented:
OK, I like the plan of including the rescaling as a second line of code in the initial pull request, and seeing if we can take these out in a subsequent one.  We can discuss whether we want to the case2 modifications to be the default today.  I actually lean a bit toward making the default "false" since this is the case that is most consistent with the physics.

Commit:
Use scale argument in get_param

Where default values were rescaled by time (sperd or spery), nutrients
(carbon to nitrogen), concentration (micromoles to moles) or light, the
optional argument scale is now set in get_param so that the values in
the parameter doc files are consistent with the literature.

theresa-cordero commented on generic_tracers/cobalt_types.F90:
```diff
@@ -55,10 +55,16 @@ module cobalt_types
+  real, parameter, public :: micromol2mol = 1.0e6     !< convert micromoles to moles (used in concentration parameters)
+  real, parameter, public :: c2n = 106.0/16.0         !< convert nutrient ratios relative to carbon to ratios relative to nitrogen
+  real, parameter, public :: micromolQpersec2W = 2.77e18/6.022e17 !< convert photosynthetically available radiation (micromole quanta
+                                                      !! per-second) to watts. 
```
@charliestock The new scale factors are added here

charliestock replied:
Looks good!

theresa-cordero commented on generic_tracers/generic_COBALT.F90:
```diff
@@ -496,456 +508,527 @@ subroutine user_add_params
-    call g_tracer_add_param('zmld_ref', cobalt%zmld_ref, 10.0)                           ! m
-    call g_tracer_add_param('densdiff_mld', cobalt%densdiff_mld, 0.03)                   ! kg m-3
-    call g_tracer_add_param('irrad_day_thresh', cobalt%irrad_day_thresh, 1.0 )           ! watts m-2
+    call get_param(param_file, "generic_COBALT", "alpha_Di_hl", phyto(DIAZO)%alpha_hl,  "alpha_Di_hl", units="g C g chl-1 s-1", &
```
@charliestock  irradiance rescaling has been added here.

charliestock replied:
Looks good.

theresa-cordero commented on generic_tracers/generic_COBALT.F90:
```diff
@@ -496,456 +508,527 @@ subroutine user_add_params
+
+    call get_param(param_file, "generic_COBALT", "kappa_eppley",        cobalt%kappa_eppley,            "kappa_eppley",        units="deg C-1",         default= 0.063)                   ! deg C-1
+    call get_param(param_file, "generic_COBALT", "P_C_max_Di_hl", phyto(DIAZO)%P_C_max_hl, "P_C_max_Di_hl", units="day-1", &
+                   default= 0.6, scale = I_sperd ) ! s-1
```
note: where variables were divided by seconds-per-day the scale factor should be `1.0/sperd` because it is multiplied. I called this `I_sperd` for "inverse seconds-per-day" but we could have just as easily defined it as `dpers`

charliestock replied:
Both approaches sound like good ideas to me.  We can stick with I_sperd unless someone has a reason to prefer dpers.

theresa-cordero commented on generic_tracers/generic_COBALT.F90:
```diff
@@ -496,456 +508,527 @@ subroutine user_add_params
-    call g_tracer_add_param('frac_mu_stress_Di',phyto(DIAZO)%frac_mu_stress,0.25)  ! none
-    call g_tracer_add_param('frac_mu_stress_Lg',phyto(LARGE)%frac_mu_stress,0.25)  ! none
-    call g_tracer_add_param('frac_mu_stress_Md',phyto(MEDIUM)%frac_mu_stress,0.25) ! none
+    call get_param(param_file, "generic_COBALT", "agg_Sm",           phyto(SMALL)%agg,  "agg_Sm", units="day-1(mol N kg)-1", &
```
add conversion from micromoles to moles.

theresa-cordero commented on generic_tracers/generic_COBALT.F90:
```diff
@@ -496,456 +508,527 @@ subroutine user_add_params
-    call g_tracer_add_param('kfe_eq_lig_hl',cobalt%kfe_eq_lig_hl, 1.0e9)                     ! mol lig-1 kg
+    call get_param(param_file, "generic_COBALT", "felig_bkg",          cobalt%felig_bkg,           "felig_bkg",         units="mol Fe kg-1", default= 0.5e-9)                          ! mol Fe kg-1
+    call get_param(param_file, "generic_COBALT", "felig_2_don",        cobalt%felig_2_don,         "felig_2_don",       units="mol lig kg-1", default= 0.5e-3)                          ! mol lig mol N-1
+    call get_param(param_file, "generic_COBALT", "fe_2_n_sed",         cobalt%fe_2_n_sed,          "fe_2_n_sed",        units="mol Fe kg-1", default= 100.0e-5 * 106 / 16)            ! mol Fe mol N-1
```
missed a c2n here

theresa-cordero commented on generic_tracers/generic_COBALT.F90:
```diff
@@ -496,456 +508,527 @@ subroutine user_add_params
-    call g_tracer_add_param('felig_bkg', cobalt%felig_bkg, 0.5e-9)                          ! mol Fe kg-1
-    call g_tracer_add_param('felig_2_don', cobalt%felig_2_don, 0.5e-3)                          ! mol lig mol N-1
-    call g_tracer_add_param('fe_2_n_sed', cobalt%fe_2_n_sed, 100.0e-5 * 106 / 16)            ! mol Fe mol N-1
-    call g_tracer_add_param('ffe_sed_max', cobalt%ffe_sed_max, 170.0/1.0e6/sperd)            ! mol Fe m-2 s-1
```
is the 1/1.0e6 here correctly reinterpreted as 1/micromol2mol ?

theresa-cordero commented on generic_tracers/generic_COBALT.F90:
```diff
@@ -496,456 +508,527 @@ subroutine user_add_params
+    call get_param(param_file, "generic_COBALT", "o2_min_nit",       cobalt%o2_min_nit,        "o2_min_nit",       units="mol O2 kg-1   ", default= o2_min_nit )                 ! mol O2 kg-1
+    call get_param(param_file, "generic_COBALT", "kappa_remin",       cobalt%kappa_remin,      "kappa_remin",      units="deg C-1       ", default= 0.063 )                      ! deg C-1
+    call get_param(param_file, "generic_COBALT", "remin_ramp_scale", cobalt%remin_ramp_scale,  "remin_ramp_scale", units="m             ", default= 50.0 )                       ! m
+    call get_param(param_file, "generic_COBALT", "rpcaco3",          cobalt%rpcaco3,           "rpcaco3",          units="mol N mol Ca-1", default= 0.070/12.0*16.0/106.0*100.0) ! mol N mol Ca-1
```
it looks like these parameters have a 1/c2n along with some other ratios. We could change them, but I think they are fines as is with a `do_not_read` flag.

theresa-cordero commented on generic_tracers/generic_COBALT.F90:
```diff
@@ -496,456 +508,582 @@ subroutine user_add_params
+                   default= 0.57, scale = I_sperd ) ! s-1
+    call get_param(param_file, "generic_COBALT", "imax_lgz", zoo(3)%imax, "imax_lgz", units="day-1", & 
+                   default= 0.23, scale = I_sperd ) ! s-1
+    call get_param(param_file, "generic_COBALT", "ki_smz",           zoo(1)%ki,                "ki_smz",           units="mol N kg-1", default=1.25e-6)                        ! moles N kg-1
```
@charliestock @yichengt900 I updated these parameters to have unique names as YC pointed out. I used the smz, mdz, and lgz rather than 1, 2, and 3 based on the other parameters and to be more clear.

charliestock replied:
Thanks Theresa.

charliestock reviewed (approved):
Thank you for your work on this pull request Theresa.  While there is still work to do on the parameter descriptions/units and deciding which parameters we should include "do_not_read" for, this falls mainly on my plate and would be best addressed through subsequent pull requests.  Thus, once we confirm that the pull request doesn't change answers, I think we should accept it and work off that version.

I did have two question on the units.  Is the scaling factor reapplied when creating the output file such that the resulting parameters are interpretable both at the input and output stage?

Also, am I right in thinking that our intent is to list the units in their input/output units rather than the units they are converted to when the model runs? 

@yichengt900, if there is a way to tag the version before this change?  Perhaps that would help those currently running simulations to avoid any issues as we perform additional testing?

yichengt900 commented:
@theresa-morrison, unfortunately, your last commit broke the automated testing again. I will have to take a look to determine why the tests are not happy with the latest changes.

theresa-cordero commented:
@yichengt900 it looks like the variables that use "do_not_read" are being set to 0.0, I think that is probably causing the issue with the testing.

yichengt900 commented:
@theresa-morrison, thanks. I think this makes sense because I keep getting `FATAL: NaN in input field of reproducing_EFP_sum(_2d)` in the automated testing with your last commit. Also, I noticed that once you introduced `do_not_read`, it changed the baseline answers. I suggest we revert the changes related to `do_not_read` and test it again.

yichengt900 reviewed (approved):
@theresa-morrison, I reverted your last commit, and now this PR passes the automated testing. I noticed there are still unresolved comments from @charliestock, but I think we can address them in a separate PR since we already have significant changes in this one. I will approve this PR and merge it soon. I will also conduct a longer test runs using NWA12 test case.

@charliestock, once this PR is merged, you can start reviewing the parameter section. I have also created a [release tag](https://github.com/NOAA-CEFI-Regional-Ocean-Modeling/ocean_BGC/releases/tag/COBALTv3_202401) for the current stable version of COBALTv3. I recommend users continue using this release tag for their production runs until the new parameter inputs are finalized.
